modules/gha/main.go

```diff
@@ -8,228 +8,217 @@
 package main
 
 import (
-	"bufio"
-	"context"
-	"errors"
-	"fmt"
+	"encoding/json"
 	"regexp"
-	"sort"
 	"strings"
 
+	"dario.cat/mergo"
+	"github.com/iancoleman/strcase"
+	"gopkg.in/yaml.v3"
+
+	"github.com/dagger/dagger/modules/gha/api"
 	"github.com/dagger/dagger/modules/gha/internal/dagger"
-	"golang.org/x/mod/semver"
-	"mvdan.cc/sh/shell"
 )
 
-func New(
-	// Disable sending traces to Dagger Cloud
-	// +optional
-	noTraces bool,
-	// Public Dagger Cloud token, for open-source projects. DO NOT PASS YOUR PRIVATE DAGGER CLOUD TOKEN!
-	// This is for a special "public" token which can safely be shared publicly.
-	// To get one, contact support@dagger.io
-	// +optional
-	publicToken string,
-	// Dagger version to run in the Github Actions pipelines
-	// +optional
-	// +default="latest"
-	daggerVersion string,
-	// Explicitly stop the Dagger Engine after completing the pipeline
-	// +optional
-	stopEngine bool,
-	// Encode all files as JSON (which is also valid YAML)
-	// +optional
-	asJSON bool,
-	// Configure a default runner for all workflows
-	// See https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/using-self-hosted-runners-in-a-workflow
-	// +optional
-	runner []string,
-	// File extension to use for generated workflow files
-	// +optional
-	// +default=".gen.yml"
-	fileExtension string,
-	// Existing repository root, to merge existing content
-	// +optional
-	// +ignore=["!.github"]
-	repository *dagger.Directory,
-	// Default timeout for CI jobs, in minutes
-	// +optional
-	timeoutMinutes int,
-) *Gha {
-	if runner == nil {
-		runner = []string{"ubuntu-latest"}
-	}
+type Gha struct{}
 
-	return &Gha{Settings: Settings{
-		PublicToken:    publicToken,
-		NoTraces:       noTraces,
-		DaggerVersion:  daggerVersion,
-		StopEngine:     stopEngine,
-		AsJSON:         asJSON,
-		Runner:         runner,
-		FileExtension:  fileExtension,
-		Repository:     repository,
-		TimeoutMinutes: timeoutMinutes,
-	}}
-}
+type Repository struct {
+	Workflows []*Workflow
 
-type Gha struct {
-	// +private
-	Pipelines []*Pipeline
-	// Settings for this Github Actions project
-	Settings Settings
+	JobDefaults      *Job      // +private
+	WorkflowDefaults *Workflow // +private
 }
 
-type Settings struct {
-	PublicToken            string
-	DaggerVersion          string
-	NoTraces               bool
-	StopEngine             bool
-	AsJSON                 bool
-	Runner                 []string
-	PullRequestConcurrency string
-	Debug                  bool
-	FileExtension          string
-	Repository             *dagger.Directory
-	TimeoutMinutes         int
-	Permissions            Permissions
+func (gha *Gha) Repository(
+	jobDefaults *Job, // +optional
+	workflowDefaults *Workflow, // +optional
+) *Repository {
```

Comment:
*Kinda* reconsidering this in light of #8792. Having the defaults set on the `Repository` is a bit odd, I wonder if you should be setting them on the constructor.

That way you could do:

```go
dagger := dag.Gha(dag.GhaOpts{Pipeline: dag.Pipeline(dag.PipelineOpts{
	OnPushBranches: []string{"main"},
	Runner: []string{"dagger-runner-1"},
})})

depot := dag.Gha(dag.GhaOpts{Pipeline: dag.Pipeline(dag.PipelineOpts{
	OnSchedule: []string{"6 0 * * *"},
	Runner: []string{"depot-runner-1"},
})})
```

Then you just pass your `Gha` objects around to create workflows, and assemble your workflows using `Repository.WithPipeline`.

Thoughts @matipan?

Reply:
> `dag.Gha(dag.GhaOpts{Workflow:`

- Assuming `Workflow:` is the default workflow, I would prefer calling it `Defaults`.
- Slight preference for pseudo-passing a `WorkflopOpts` argument instead of a `Workflow`. I know that involves a lot of copy-pasting today, but IMO it's still worth it. Easier to change internal implementation than user-facing API.
- I can live with keeping the `Workflow` type as long as we call it `Defaults` (in other words: an argument name that corresponds to the intent).

Reply:
I agree that settings the defaults at a constructor level gives users more flexibility without introducing a new thing into the module. Having two `Repository` objects would be a bit confusing. `WorkflowGroups` would be an alternative but I hate it 😆. Constructor-based defaults sounds good!

Reply:
Okay, made these changes - `Repository` is gone :tada: Now it's much clearer and succinct.

Take a look at the new `.github/main.go` for a more complicated example: we have two `Gha` objects - one as a `Workflows`, another as a `DaggerRunner`. This complexity isn't required *right* now, they could just be the same - but it's a baseline for how we should implement https://github.com/dagger/dagger/pull/8792. The runners have defaults for triggers+github runners, and we can easily split out different jobs to different jobs. This is possible because we take into account *both* the `Gha`s: all the defaults always propagate, no matter where they're specified.